src/api/services/index.ts

```diff
@@ -0,0 +1,111 @@
+import { Client } from './client';
+import { Identity } from '../../identity';
+import { Provider } from './provider';
+import { Base } from '../base';
+import Transport, { Message, Payload } from '../../transport/transport';
+
+export interface Options {
+    wait?: boolean;
+    uuid: string;
+    payload?: any;
+}
+
+export interface ServicePayload {
+    payload: Payload;
+}
+export interface ServiceMessage extends Message<any> {
+  senderIdentity: Identity;
+  ackToSender: any;
+  serviceIdentity: Identity;
+  connectAction: boolean;
+}
+
+export class Service extends Base {
+    private serviceMap: Map<string, Provider | Client>;
+    constructor(wire: Transport) {
+        super(wire);
+        this.serviceMap = new Map();
+        wire.registerMessageHandler(this.onmessage.bind(this));
+    }
+
+    public async onServiceConnect(identity: Identity, listener: EventListener): Promise<void> {
+            this.registerEventListener({
+                topic: 'service',
+                type: 'connected',
+                ...identity
+            });
+            this.on('connected', listener);
+    }
+
+    public async connect(options: Options): Promise<Client> {
+        try {
+            const { payload: { data: serviceIdentity } } = await this.wire.sendAction('send-service-message', Object.assign({
+                connectAction: true,
+                wait: true
+            }, options));
+            const channel = new Client(serviceIdentity, this.wire.sendAction.bind(this.wire));
+            channel.onServiceDisconnect = (listener: () => void) => {
+                this.registerEventListener({
+                    topic: 'service',
+                    type: 'disconnected',
+                    ...serviceIdentity
+                });
+                this.on('disconnected', listener);
+            };
+            this.serviceMap.set(serviceIdentity.uuid, channel);
+            return channel;
+        } catch (e) {
+            throw new Error(e.message);
```

Comment:
Two questions: 
1) Why not `throw e`?

2) if we are not handling the error in any meaningful way, why catch at all ?

Reply:
What we are catching is a nack from the core which is coming through as a RuntimeError, when this could very well be an intentional error from the user (ie authenticated Service). I'm catching it an throwing it back as a regular error because in those cases it is not a runtime error.